### PR TITLE
Post Content: Fix display of block support styles

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes:
- https://github.com/WordPress/gutenberg/issues/65644

Revisits approach from:
- https://github.com/WordPress/gutenberg/pull/35863

## What?

Renders actual blocks for Post Content previews when the user can edit the post being previewed. This allows the editors block support hooks to do their job and render additional styles as needed.

Props to @andrewserong for the approach.

## Why?

Currently, the Post Content block renders read-only content via a REST API call. The server side rendered content returned via that request lacks additional stylesheets normally enqueued by block supports. This results in element, layout, and block style variation styles all missing when viewing Query loops containing Post Content previews.

## How?

https://github.com/WordPress/gutenberg/pull/35863 was originally blocked due to concern around potentially exposing private block data if the raw block data was rendered. This PR takes a ever-so-slightly different approach in that it will only render the raw blocks instead of server side rendered HTML when the user has access to edit the post in question.

My understanding is that plugins providing visibility control over blocks within a page, still allow any user with rights to edit a particular page to edit those visibility or access-controller blocks e.g. change them, alter the user capabilities required for their display etc. In other words, the access control there is primarily for the frontend. (happy to be corrected here 🙂)

If a post is password protected and a user doesn't have edit permissions for that post, the original warning is still displayed in place of the post's content.

## Testing Instructions

Using TT5:

1. Create a post and add some content including a nested block with custom layout e.g. Row or Grid block
2. Select a container block and apply a section style provided via TT5
3. Add a paragraph containing a link to the content and apply a custom link color to that block
4. Save and publish the post
5. Switch to the Site Editor and edit the home template
6. Confirm that the query loop in that template renders the post content appropriately.

<details>
<summary>Example block content to test with if it helps</summary>

```html
<!-- wp:group {"className":"is-style-section-4","style":{"spacing":{"padding":{"right":"var:preset|spacing|40","left":"var:preset|spacing|40","top":"var:preset|spacing|40","bottom":"var:preset|spacing|40"}}},"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group is-style-section-4" style="padding-top:var(--wp--preset--spacing--40);padding-right:var(--wp--preset--spacing--40);padding-bottom:var(--wp--preset--spacing--40);padding-left:var(--wp--preset--spacing--40)"><!-- wp:group {"style":{"layout":{"columnSpan":1,"rowSpan":1}},"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#29c8f8"}}}}} -->
<p class="has-link-color">Welcome to <a href="http://www.wordpress.org">WordPress</a>. This is your first post. Edit or delete it, then start writing! Welcome to WordPress. </p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"style":{"elements":{"link":{"color":{"text":"#3ee732"}}}}} -->
<p class="has-link-color">Welcome to WordPress. This is your first post. Edit or delete it, then start writing! Welcome to <a href="http://wordpress.org">WordPress</a>. </p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->

<!-- wp:buttons {"style":{"layout":{"columnSpan":2,"rowSpan":1}},"layout":{"type":"flex","justifyContent":"center"}} -->
<div class="wp-block-buttons"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">Button</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons --></div>
<!-- /wp:group -->

<!-- wp:image {"sizeSlug":"full","linkDestination":"none"} -->
<figure class="wp-block-image size-full"><img/></figure>
<!-- /wp:image --></div>
<!-- /wp:group -->
```
</details>

## Screenshots or screencast 

| Post | Before | After |
|---|---|---|
| Screenshot 2024-10-09 at 6 41 33 pm | Screenshot 2024-10-09 at 6 41 17 pm | Screenshot 2024-10-10 at 3 00 10 pm |

### CODE REVIEW

**Review by andrewserong — COMMENTED** [review]
I really like the approach here, I think it's a good pragmatic path forward for 6.7, especially since the Post Content block is used so prominently there. I tested this very quickly in TT4, too, with element styles and it displays block content consistently for users who can edit the post that's being previewed:

image

Overall, especially since most of the time users who are interacting with the post content block will have access to edit the posts that they're previewing, I think this fix should cover most cases.

Let's see what others think!

**Comment by jasmussen** [comment]
This looks great, thank you for working on this, and as shown below, it solves the issue for me:

branch

Note though that I'm seeing here an old issue pop up again, [namely that full-wide content becomes content-wide in the TT5 Blog Home template](https://github.com/WordPress/twentytwentyfive/issues/428). This was supposedly [fixed here](https://github.com/WordPress/twentytwentyfive/pull/455). So I dove in and found that, unfortunately that appears to be working in trunk Gutenberg, meaning wide and fullwide may have broken in this PR:

trunk

So, section styles are working, but wide/fullwide broke. Is the latter possible to fix in this one?

**Comment by carolinan** [comment]
Yes unfortunately I came here right now after testing to report the same.
In the Site Editor the section style colors are showing but wide and full width blocks are now content width.

The "Instagram grid" pattern with the group block grid variation that is broken on trunk Gutenberg and on Beta 2 is now using the grid, but at content width: 
image

**Comment by aaronrobertshaw** [comment]
Thanks for the testing and feedback @jasmussen and @carolinan 👍 

I suspect this might be related to how the [layout classes are applied](https://github.com/WordPress/gutenberg/pull/66003/files#diff-370d270911017533256dc539d4a3bb75b9fb15bf9b21ff874c43517b3b8877fbR39) to the wrapper around the read-only post content while the inner content has a specific layout. I'm out of time today but will see what can be done here first thing tomorrow.

**Comment by ndiego** [comment]
Just reiterating @jasmussen and @carolinan's feedback above. 

| Post Editor | Site Editor | Front end |
|-|-|-|
|image|image|image|

**Comment by ntsekouras** [comment]
👋 - I pushed something and would be great if you could retest. Thank you!

**Comment by colorful-tones** [comment]
> 👋 - I pushed something and would be great if you could retest. Thank you!

This seems to be working!

https://github.com/user-attachments/assets/1a533a94-d95e-4954-8ab8-78eaaa932eae

**Comment by peterwilsoncc** [comment]
Based on some quick testing, I think this could work as a temporary fix. It would be better to get the styles available to the assets when using `content.rendered` but the most likely scenario is that someone with theme edit permissions has edit_others_posts permissions so I think it works for the site editor.

I'm about to step out for an appointment so will come back to this in a few hours to follow up further.

**Comment by ramonjd** [comment]
LGTM - agree that this is a huge improvement and a good fix to get in now.




## Before


https://github.com/user-attachments/assets/ab7b4fc7-94a3-4241-98d7-969a0b3e12f3


## After

https://github.com/user-attachments/assets/a4f207cc-49a3-429f-8e9c-46229831eba8

**Review by ramonjd — APPROVED** [review]

**Comment by aaronrobertshaw** [comment]
Thanks for pushing the parent layout fix @ntsekouras 🙇 It makes sense and tests well for me also.

I appreciate everyone jumping in to test and review this one. I'll hold off for a little longer before merging to give @peterwilsoncc a chance to take a closer look.

>  It would be better to get the styles available to the assets when using content.rendered

Agreed. The thinking was however that approach might be a much larger change with broader impact. At this stage of the release cycle it felt like a bridge too far for 6.7.

**Review by peterwilsoncc — APPROVED** [review]
These changes look good to me from a functionality perspective. 

I used a very simple filter to confirm the raw data is only used for users with edit permissions and it behaved as expected.

```php
add_filter( 'map_meta_cap', function( $caps, $cap, $user_id, $args ) {
	if ( 'edit_post' === $cap || 'delete_post' === $cap ) {
		$caps = array( 'do_not_allow' );
	}

	return $caps;
}, 10, 4 );
```

I also confirmed that the additional use of `userCanEdit` doesn't result in additional HTTP requests and it does not. 

As a few of you have approved the code I think this is good to merge.

**Review by andrewserong — APPROVED** [review]
LGTM, too, thanks for honing in on a pragmatic fix for 6.7 @aaronrobertshaw! 🙇

**Comment by aaronrobertshaw** [comment]
Nice teamwork on this one, thanks everyone 👍 

I'll get this merged then!

**Comment by jasmussen** [comment]
Kudos, thank you all for this work and the reviews! This is impactful.

**Pull request merged**
